site2/docs/io-develop.md

```diff
@@ -1,156 +1,188 @@
-Pulsar IO runtime.
+    Besides, Pulsar runtime also provides a `SourceContext` for the 
+    connector to access runtime resources for tasks like collecting metrics. The implementation can save the `SourceContext` for future use.
 
-These information includes:
+2. Implement the {@inject: github:`read`:/pulsar-io/core/src/main/java/org/apache/pulsar/io/core/Source.java#L41} method.
 
-- *Topic Name*: _Optional_. If the record is originated from a Pulsar topic, it should be the Pulsar topic name.
-- *Key*: _Optional_. If the record has a key associated with it.
-- *Value*: _Required_. The actual data of this record.
-- *Partition Id*: _Optional_. If the record is originated from a partitioned source,
-  return its partition id. The partition id will be used as part of the unique identifier
-  by Pulsar IO runtime to do message deduplication and achieve exactly-once processing guarantee.
-- *Record Sequence*: _Optional_. If the record is originated from a sequential source,
-  return its record sequence. The record sequence will be used as part of the unique identifier
-  by Pulsar IO runtime to do message deduplication and achieve exactly-once processing guarantee.
-- *Properties*: _Optional_. If the record carries user-defined properties, return those properties.
+    ```java
+        /**
+        * Reads the next message from source.
+        * If source does not have any new messages, this call should block.
+        * @return next message from source.  The return result should never be null
+        * @throws Exception
+        */
+        Record<T> read() throws Exception;
+    ```
 
-Additionally, the implemention of the record should provide two methods: `ack` and `fail`. These
-two methods will be used by Pulsar IO connector to acknowledge the records that it has done
-processing and fail the records that it has failed to process.
+    If nothing to return, the implementation should be blocking rather than returning `null`. 
 
-{@inject: github:`KafkaSource`:/pulsar-io/kafka/src/main/java/org/apache/pulsar/io/kafka/KafkaAbstractSource.java} is a good example to follow.
+    The returned {@inject: github:`Record`:/pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java#L28} should encapsulate the following information, which is needed by Pulsar IO runtime. 
 
-#### Develop a sink connector
+    * {@inject: github:`Record`:/pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java#L28} should provide the following variables:
 
-Developing a sink connector is as easy as developing a source connector. You just need to
-implement {@inject: github:`Sink`:/pulsar-io/core/src/main/java/org/apache/pulsar/io/core/Sink.java} interface.
+      |Variable|Required|Description
+      |---|---|---
+      `TopicName`|No|Pulsar topic name from which the record is originated from.
+      `Key`|No| |
```

Comment:
@tuteng It seems that the description of `Key` in [`Record.java`](https://github.com/apache/pulsar/blob/master/pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java#L28) is incorrect.

Could you please provide a description for `Key`?

Reply:
`Key` is related to the routing mode of the message http://pulsar.apache.org/docs/en/concepts-messaging/#routing-modes